lens.swift

```diff
@@ -0,0 +1,132 @@
+import Foundation
+import AppKit
+import CoreImage
+
+// Persists data to the filesystem sans any wrappers.
+// Probably should look into using UserDefaults or FileManager.
+// See: https://programmingwithswift.com/save-images-locally-with-swift-5/
+func save(data:Data, path:String) {
+  print("PATH \(path)")
+  guard let url = URL(string: path) else {
+    print("Bad URL")
+    return
+  }
+  print("Writing \(url)")
+  try! data.write(to: url)
+  print("Wrote \(url)")
+}
+
+// Retrieves data from a string and passes it to the callback
+func download(urlString: String, callback: @escaping (Data) -> ()) {
+  guard let url = URL(string: urlString) else {
+    print("Bad URL")
+    return
+  }
+  print("A")
+  let task = URLSession.shared.dataTask(with: url) { data, response, error in
+    print("B")
+    guard let data = data, error == nil else {
+      print("OH NO")
+      return
+    }
+    print("C")
+
+    print(data)
+    callback(data)
+    print("D")
+  }
+  // tasks start paused; so we need to start them
+  task.resume()
+  // And because we are in a swift command line program,
+  // we're going to sleep so that the exectuion doesn't stop
+  // before the file is downloaded
+  while !task.progress.isFinished {
+    print(".")
+    sleep(1)
+  }
+}
+
+
+func saveCIImage(data:CIImage, path:String) {
+  print("PATH - saveCIImage: \(path)")
+  guard let url = URL(string: path) else {
+    print("Bad URL - saveCIImage")
+    return
+  }
+  let rep = NSBitmapImageRep(ciImage: data)
+  let PNGData = rep.representation(using: NSBitmapImageRep.FileType.png, properties:[:])
+  print("Writing - saveCIImage \(url)")
+  try! PNGData!.write(to: url);
+  print("Wrote - saveCIImage \(url)")
+}
+
+
+
+class Filter {
+    let filterName: String
+    var filterData: Dictionary<String, Any>?
+
+    init(filterName: String, filterData: Dictionary<String, Any>?) {
+        self.filterName = filterName
+        self.filterData = filterData
+    }
+}
+
+// A Fingerprint is a random-number generator that is used by Lenses
+// to adjust the image.
+class Fingerprint {
+
+}
+
+// A Lens applied to a CGImage returns a new CGImage the filter transformations applied
+// currently only works with CIImage
+class Lens {
+  var fingerprint: Fingerprint
+  var filter: Filter
+  init(fingerprint:Fingerprint, filter:Filter) {
+    self.fingerprint = fingerprint
+    self.filter = filter
+  }
+
+  func apply(image: CGImage) -> CGImage {
+    // let cgimg = context.createCGImage(outputImage, from: outputImage.extent)
+    return image
+  }
+
+  func apply(image: CIImage) -> CIImage {
+    let localFilter = CIFilter(name: filter.filterName, parameters: filter.filterData)
+    localFilter?.setValue(image, forKey: kCIInputImageKey)
```

Comment:
This looks great! It feels a little bit like [feature envy](https://refactoring.guru/smells/feature-envy) for the Lens to be pulling this data out of the filter. I wonder if we can apply ["Tell, Don't Ask"](https://www.martinfowler.com/bliki/TellDontAsk.html) here and refactor in such a way that the Lens method takes responsibility for handling the failure case for Filters (i.e. the filter raises, returns nil, etc.).

```
guard let output = filter?.apply(image, fingerprint) else {
  return image
}
return output
```

This way each filter can encapsulate how it leverages the fingerprint to generate the filter parameters. 

For example, a MotionBlur filter looks like it wants two numbers between ??? and 20?, so the MotionBlur filter class would use the Fingerprint to generate two random values. On the other hand, the SepiFilter class would only need to generate a single float value; so it would use the fingerprint to do that. Her'es some pseudocode for those two filters:

```
class SepiaFilter  {
  func apply(image, fingerprint) {
    intensity = fingerprint.randBetween(0.8,1.0)
    let localFilter = CIFilter(name: "CISepiaTone", parameters: [kCIInputIntensityKey: intensity])
    localFilter?.setValue(image, forKey: kCIInputImageKey)
    return localFilter.?outputImage
  }
}
```

Once we have 2~3 classes, we may find a way to DRY it up a bit.
```